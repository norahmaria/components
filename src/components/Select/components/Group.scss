@use '../../../scss/variables' as *;

.select-nm > .options-nm > .group-nm + .group-nm,
.select-nm > .options-nm > .group-nm ~ .group-nm {
  ul {
    border-bottom: none;
    margin-bottom: 0;
    padding-bottom: 0;
  }
}

.select-nm > .options-nm > .group-nm:first-of-type {
  ul {
    padding-top: 1.25rem;
  }
}

.select-nm > .options-nn .option-nm + .group-nm {
  margin-top: 0.75rem;
}

.select-nm > .options-nm > .group-nm {
  ul {
    display: grid;
    gap: 0.35rem;
    width: 100%;

    .title-nm {
      display: flex;
      gap: 0.3rem;
      align-items: center;

      font-size: 0.875em;
      padding: 0 0.5em 0.5em 0.5rem;
      overflow: hidden;

      text-transform: uppercase;
      font-weight: 600;

      color: inherit;
      color: rgba($black, 0.65);

      svg {
        transform: rotate(180deg);

        path,
        rect {
          fill: rgba($black, 0.45);
        }
      }
    }

    .option-nm {
      &:last-of-type {
        padding-bottom: 0.25rem;
      }
    }
  }
}
