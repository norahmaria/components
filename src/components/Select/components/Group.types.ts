import React from 'react'

interface GroupProps {
  title: string
  children: React.ReactChild | React.ReactChild[]
}

export default GroupProps
