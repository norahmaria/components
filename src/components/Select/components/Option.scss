@use '../../../scss/variables' as *;

.select-nm > .options > .option,
.select-nm > .options > .group .option {
  display: flex;
  justify-content: space-between;
  align-items: center;
  list-style: none;
  gap: 0.5rem;

  border: 1px solid transparent;
  border-radius: $br;

  padding: 0.4em 0.5em;

  transition: background 0.25s ease-in-out, color 0.25s ease-in-out,
    border 0.25s ease-in-out, padding 0.25s ease-in-out;
  cursor: pointer;

  .label {
    width: 100%;
    display: flex;
    align-items: center;
    gap: 0.5rem;
  }

  &,
  &[aria-selected='true'] {
    &:hover,
    &:focus-visible,
    &:active {
      outline: none;
      background: darken($white, 5%);
      border: 1px solid lighten($black, 75.5%);
      padding-left: 1em;
    }

    &:focus-visible {
      background: lighten($black, 77.5%);
    }
  }

  &[aria-disabled='true'] {
    background: darken($white, 3%);
    color: #898a8a;
    cursor: not-allowed;

    &:hover {
      padding: 0.4em 0.5em;
    }

    svg {
      opacity: 0.45;
      filter: grayscale(100%);
    }
  }

  &[aria-selected='true'] {
    font-weight: 500;
  }

  &:first-child {
    margin-top: 0.9rem;
  }

  &:last-child {
    margin-bottom: 0.9rem;
  }
}

@each $name, $color in $colors {
  @if $name == 'neutral' {
    .select-nm.#{$name} > .options > .option,
    .select-nm.#{$name} > .options > .group .option {
      &[aria-selected='true'] {
        background: scale-color($color, $saturation: 5%, $lightness: 50%);
        border: 1px solid scale-color($color, $saturation: 5%, $lightness: 50%);
      }
    }
  } @else {
    .select-nm.#{$name} > .options > .option,
    .select-nm.#{$name} > .options > .group .option {
      &[aria-selected='true'] {
        background: scale-color($color, $saturation: 5%, $lightness: 90%);
        border: 1px solid scale-color($color, $saturation: 5%, $lightness: 90%);
      }
    }
  }
}
