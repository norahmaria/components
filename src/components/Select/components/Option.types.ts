interface OptionProps {
  value: string | number
  leftIcon?: JSX.Element | string
  rightIcon?: JSX.Element | string
  disabled?: boolean
  children: string
}

export default OptionProps
