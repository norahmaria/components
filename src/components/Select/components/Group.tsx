import React from 'react'
import './Group.scss'

import { ReactComponent as LineArrow } from '../../../assets/Arrow_Line.svg'
import GroupProps from '../components/Group.types'

const Group: React.FC<GroupProps> = ({ title, ...props }) => {
  return (
    <ul className="group" {...props}>
      <div className="title">
        {/* <LineArrow /> */}
        {title}
      </div>
      {props.children}
    </ul>
  )
}

export default Group
