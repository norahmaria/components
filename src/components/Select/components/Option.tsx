import React from 'react'
import './Option.scss'

import OptionProps from './Option.types'

const Option = ({}: OptionProps) => <></>

export default Option
