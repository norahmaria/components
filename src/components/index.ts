export { default as Button } from './Button/Button'
export { default as Checkbox } from './Checkbox/Checkbox'

import SelectComponent from './Select/Select'
import Option from './Select/components/Option'
import Group from './Select/components/Group'

import SelectProps from './Select/Select.types'
import OptionProps from './Select/components/Option.types'
import GroupProps from './Select/components/Group.types'

type SelectWithChildren = React.FC<SelectProps> & {
  Option: React.FC<OptionProps>
  Group: React.FC<GroupProps>
}

const Select: SelectWithChildren = SelectComponent as SelectWithChildren
Select.Option = Option
Select.Group = Group

export { Select }
