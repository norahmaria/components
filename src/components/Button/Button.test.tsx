import React from 'react'
import { act, render, waitFor } from '@testing-library/react'

import { Button } from '../../components/index'

describe('Button Component', () => {
  it('Can render', () => {
    const { getByRole } = render(<Button onClick={() => {}}>Hello World</Button>)
    expect(getByRole('button')).toMatchSnapshot()
  })

  it('Can render non default variants', () => {
    const buttons = render(
      <>
        <Button onClick={() => {}} variant="outlined">
          Hello World
        </Button>
        <Button onClick={() => {}} variant="secondary">
          Hello World
        </Button>
      </>
    )

    expect(buttons.getAllByRole('button')).toMatchSnapshot()
  })

  it('Can toggle click class', async () => {
    const { getByRole } = render(<Button onClick={() => {}}>Hello World</Button>)
    const button = getByRole('button')

    button.click()
    expect(button).toHaveClass('clicked-true')

    await waitFor(() => {
      expect(button).not.toHaveClass('clicked-true')
    })
  })

  it('Can render loading icon', () => {
    const { getByTestId } = render(
      <Button onClick={() => {}} isLoading={true}>
        Hello World
      </Button>
    )
    const loadingIcon = getByTestId('btn-loading-icon')

    expect(loadingIcon).toBeInTheDocument()
  })

  it('Ignores click if button is loading or disabled', () => {
    const loadingClick = jest.fn()
    const disabledClick = jest.fn()

    const { getByTestId, getByText } = render(
      <>
        <Button onClick={jest.fn()} isLoading={true}>
          Loading Button
        </Button>
        <Button onClick={jest.fn()}>Disabled Button</Button>
      </>
    )

    const loadingButton = getByTestId('btn-loading-icon')
    const disabledButton = getByText('Disabled Button')

    act(() => {
      loadingButton.dispatchEvent(new MouseEvent('click', { bubbles: true }))
      disabledButton.dispatchEvent(new MouseEvent('click', { bubbles: true }))
    })

    expect(loadingClick).not.toHaveBeenCalled()
    expect(disabledClick).not.toHaveBeenCalled()
  })
})
