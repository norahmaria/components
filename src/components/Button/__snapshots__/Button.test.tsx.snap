// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Can render 1`] = `
<button
  class="
        btn-nm
        size-medium
        color-primary
        variant-default
        clicked-false
        
        
        undefined
      "
  type="button"
>
  Hello World
</button>
`;

exports[`Button Component Can render non default variants 1`] = `
Array [
  <button
    class="
        btn-nm
        size-medium
        color-primary
        variant-outlined
        clicked-false
        
        
        undefined
      "
    type="button"
  >
    Hello World
  </button>,
  <button
    class="
        btn-nm
        size-medium
        color-primary
        variant-secondary
        clicked-false
        
        
        undefined
      "
    type="button"
  >
    Hello World
  </button>,
]
`;
